From e81d56d247a8c93d24e03be378d1748f3e044e6b Mon Sep 17 00:00:00 2001
From: Ladislav Michl <ladis@linux-mips.org>
Date: Tue, 9 Jan 2018 14:19:11 +0100
Subject: [PATCH] VGV952CJW33-E-IR: mtd: nand: samsung: Disable subpage writes on 21nm NAND.

Some Samsung SLC NAND are manufactured using the 21nm process.
They does not supports partial page programming, so disable subpage writes
for it. Manufacturing process is stored in lowest two bits of 5th ID
byte.

This patch is derived and adapted from the upstream patch which
handles a different samsung NAND flash devie (K9F1G08U0E) and is named
mtd: nand: samsung: Disable subpage writes on E-die NAND
and available since kernel release v4.16.
---
 drivers/mtd/nand/nand_samsung.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

Index: linux-5.4.46/drivers/mtd/nand/raw/nand_samsung.c
===================================================================
--- linux-5.4.46.orig/drivers/mtd/nand/raw/nand_samsung.c
+++ linux-5.4.46/drivers/mtd/nand/raw/nand_samsung.c
@@ -12,6 +12,7 @@ static void samsung_nand_decode_id(struc
 {
 	struct mtd_info *mtd = nand_to_mtd(chip);
 	struct nand_memory_organization *memorg;
+	u8 *d  = chip->id.data;
 
 	memorg = nanddev_get_memorg(&chip->base);
 
@@ -92,13 +93,19 @@ static void samsung_nand_decode_id(struc
 		}
 	} else {
 		nand_decode_ext_id(chip);
-
 		if (nand_is_slc(chip)) {
 			switch (chip->id.data[1]) {
 			/* K9F4G08U0D-S[I|C]B0(T00) */
 			case 0xDC:
 				chip->base.eccreq.step_size = 512;
 				chip->base.eccreq.strength = 1;
+				if (chip->id.len > 4 &&
+				    (chip->id.data[4] & GENMASK(1, 0)) == 0x1) {
+					chip->options |= NAND_NO_SUBPAGE_WRITE;
+					pr_debug("samsung_nand_decode_id: id.data[1] is 0x%02X, disabling subpage writes\n", d[1]);
+				} else {
+					pr_debug("samsung_nand_decode_id: id.data[1] is 0x%02X, allowing subpage writes\n", d[1]);
+				}
 				break;
 
 			/* K9F1G08U0E 21nm chips do not support subpage write */
@@ -117,6 +124,9 @@ static void samsung_nand_decode_id(struc
 static int samsung_nand_init(struct nand_chip *chip)
 {
 	struct mtd_info *mtd = nand_to_mtd(chip);
+	u8 *d  = chip->id.data;
+	pr_debug("samsung_nand_decode_id: ID is len=%d, %02X %02X %02X %02X %02X %02X %02X %02X\n",
+		chip->id.len, d[0], d[1], d[2], d[3], d[4], d[5], d[6], d[7]);
 
 	if (mtd->writesize > 512)
 		chip->options |= NAND_SAMSUNG_LP_OPTIONS;
